### PR TITLE
Switch to using kubelet config file for all supported flags

### PR DESCRIPTION
**Pull request by brandond**
#### Proposed Changes ####

Switch to using kubelet config file for all supported flags

#### Types of Changes ####

enhancement / tech debt

#### Verification ####

See linked issue

#### Linked Issues ####

* https://github.com/k3s-io/k3s/issues/2116

#### User-Facing Change ####

```release-note
```

#### Further Comments ####

Currently using a hacky string replace when marshaling the config file to work around an upstream issue:
* https://github.com/kubernetes/kubernetes/issues/79421

### CODE REVIEW

**Review by manuelbuil — DISMISSED** [review]

File: pkg/daemons/agent/agent_windows.go
Comment (dereknola) [inline pkg/daemons/agent/agent_windows.go:46]: Is this going to work on Windows? Won't stuff like the cgroup driver fail/be unnecessary on this OS?

Comment (brandond) [reply pkg/daemons/agent/agent_windows.go:46]: yeah I'm honestly not sure what the windows kubelet does if you try to set the cgroup stuff. I'll try this out in RKE2 before merging.

**Review by dereknola — APPROVED** [review]
LGTM pending RKE2 cgroup checks

**Comment by brandond** [comment]
I'm going to hold on this until the August release cycle, as I'd like more lead time on testing it.

**Comment by brandond** [comment]
@dereknola I have confirmed this works on Windows, using RKE2: https://github.com/rancher/rke2/pull/6909

```console
root@dev02:~# kubectl get node -o wide
NAME    STATUS   ROLES                       AGE     VERSION          INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                                  KERNEL-VERSION     CONTAINER-RUNTIME
dev02   Ready    control-plane,etcd,master   3h42m   v1.31.1+rke2r1   10.0.1.184    <none>        Ubuntu 24.04 LTS                          6.8.0-45-generic   containerd://1.7.21-k3s2
win02   Ready    <none>                      3h31m   v1.31.1          10.0.1.224    <none>        Windows Server 2022 Standard Evaluation   10.0.20348.2113    containerd://1.7.21-k3s2
```

```
PS C:\Users\Administrator> c:\usr\local\bin\rke2.exe --version
rke2.exe version v1.31.1+dev.045680fc (045680fc54b2e82cbb5491c9a916655535cd8fb7)
go version go1.22.6

PS C:\Users\Administrator> dir C:\var\lib\rancher\rke2\agent\etc\kubelet.conf.d\


    Directory: C:\var\lib\rancher\rke2\agent\etc\kubelet.conf.d


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
-a----         10/5/2024  12:51 AM           1516 00-rke2-defaults.conf


PS C:\Users\Administrator> cat C:\var\lib\rancher\rke2\agent\etc\kubelet.conf.d\00-rke2-defaults.conf
address: 0.0.0.0
apiVersion: kubelet.config.k8s.io/v1beta1
authentication:
  anonymous:
    enabled: false
  webhook:
    cacheTTL: 2m0s
    enabled: true
  x509:
    clientCAFile: C:\var\lib\rancher\rke2\agent\client-ca.crt
authorization:
  mode: Webhook
  webhook:
    cacheAuthorizedTTL: 5m0s
    cacheUnauthorizedTTL: 30s
cgroupDriver: cgroupfs
clusterDNS:
- 10.43.0.10
clusterDomain: cluster.local
containerRuntimeEndpoint: npipe:////./pipe/containerd-containerd
cpuManagerReconcilePeriod: 10s
evictionHard:
  imagefs.available: 5%
  nodefs.available: 5%
evictionMinimumReclaim:
  imagefs.available: 10%
  nodefs.available: 10%
evictionPressureTransitionPeriod: 5m0s
failSwapOn: false
fileCheckFrequency: 20s
healthzBindAddress: 127.0.0.1
httpCheckFrequency: 20s
imageMaximumGCAge: 0s
imageMinimumGCAge: 2m0s
kind: KubeletConfiguration
logging:
  flushFrequency: 5s
  format: text
  options:
    json:
      infoBufferSize: "0"
    text:
      infoBufferSize: "0"
  verbosity: 0
memorySwap: {}
nodeStatusReportFrequency: 5m0s
nodeStatusUpdateFrequency: 10s
resolverConfig: C:\var\lib\rancher\rke2\agent\etc\resolv.conf
runtimeRequestTimeout: 2m0s
serializeImagePulls: false
shutdownGracePeriod: 0s
shutdownGracePeriodCriticalPods: 0s
staticPodPath: C:\var\lib\rancher\rke2\agent\pod-manifests
streamingConnectionIdleTimeout: 4h0m0s
syncFrequency: 1m0s
tlsCertFile: C:\var\lib\rancher\rke2\agent\serving-kubelet.crt
tlsPrivateKeyFile: C:\var\lib\rancher\rke2\agent\serving-kubelet.key
volumeStatsAggPeriod: 1m0s
```

**Review by galal-hussein — DISMISSED** [review]

**Comment by brandond** [comment]
moving to WIP until closer to 1.32.

**Comment by caroline-suse-rancher** [comment]
Hey @brandond, are we still planning to do this for 1.32?

**Comment by brandond** [comment]
I am watching what upstream does with this feature. It looks like it might be another minor release until it is on by default.

File: pkg/daemons/agent/agent.go
Comment (dereknola) [inline pkg/daemons/agent/agent.go:161]: Does this need to be changed now that 1.32 is here?

Comment (brandond) [reply pkg/daemons/agent/agent.go]: dropped.

Comment (brandond) [reply pkg/daemons/agent/agent.go]: ugh nope, still required
```console
root@systemd-node-1:/# grep resolv /var/lib/rancher/k3s/agent/etc/kubelet.conf.d/00-k3s-defaults.conf
resolvConf: /var/lib/rancher/k3s/agent/etc/resolv.conf

root@systemd-node-1:/# kubectl get --raw /api/v1/nodes/systemd-node-1/proxy/configz 2>/dev/null | jq . | grep resolv
    "resolvConf": "/etc/resolv.conf",

root@systemd-node-1:/# echo -e "apiVersion: kubelet.config.k8s.io/v1beta1\nkind: KubeletConfiguration\nresolverConfig: /var/lib/rancher/k3s/agent/etc/resolv.conf" > /var/lib/rancher/k3s/agent/etc/kubelet.conf.d/99-resolv.conf

root@systemd-node-1:/# systemctl restart k3s

root@systemd-node-1:/# grep -rF resolv /var/lib/rancher/k3s/agent/etc/kubelet.conf.d/
/var/lib/rancher/k3s/agent/etc/kubelet.conf.d/00-k3s-defaults.conf:resolvConf: /var/lib/rancher/k3s/agent/etc/resolv.conf
/var/lib/rancher/k3s/agent/etc/kubelet.conf.d/99-resolv.conf:resolverConfig: /var/lib/rancher/k3s/agent/etc/resolv.conf

root@systemd-node-1:/# kubectl get --raw /api/v1/nodes/systemd-node-1/proxy/configz 2>/dev/null | jq . | grep resolv
    "resolvConf": "/var/lib/rancher/k3s/agent/etc/resolv.conf",
```

**Review by dereknola — APPROVED** [review]

**Review by dereknola — APPROVED** [review]

**Review by galal-hussein — APPROVED** [review]

**Pull request merged**
